File: Nick-Haskell/PokerPuzzle.hsproj/Parsing.hs
{-# LANGUAGE OverloadedStrings #-}

module Parsing where

import Data.Text.Lazy (Text, split, unpack)
import qualified Data.Text.Lazy as T (take, drop) 
import Model

parseLine :: Text -> (Hand, Hand)
parseLine line =
  let
    cardsText = split (==' ') line
    cards = map textToCard cardsText
  in
    (Hand $ take 5 cards, Hand $ drop 5 cards)
    

textToCard :: Text -> Card
textToCard cardText = 
  let
    valueText = T.take 1 cardText
    suitText = T.drop 1 cardText
  in
    Card (parseValue valueText) (parseSuit suitText)
    
parseValue :: Text -> Value
parseValue "T" = Ten
parseValue "A" = Ace
parseValue "K" = King
parseValue "Q" = Queen
parseValue "J" = Jack
parseValue valueText = toEnum $ (read $ unpack valueText) - 2 
    
parseSuit :: Text -> Suit
parseSuit "C" = Club
parseSuit "S" = Spade
parseSuit "H" = Heart
parseSuit "D" = Diamond
    

File: Nick-Swift/PokerPuzzle/Parsing.swift
import Foundation

func parseLine(line: String) -> (Hand, Hand) {
    let cardsText = line.components(separatedBy: " ")
    let cards = cardsText.map(textToCard)
    
    return (Array(cards.dropLast(5)), Array(cards.dropFirst(5)))
}

func textToCard(cardText:String) -> Card {
    let valueText = String(cardText.characters.prefix(1))
    let suitText = String(cardText.characters.suffix(1))
    
    return Card(value: parseValue(valueText), suit: parseSuit(suitText))
}

func parseValue(_ valueText:String) -> Value {
    switch valueText {
    case "T":
        return .ten
    case "A":
        return .ace
    case "K":
        return .king
    case "Q":
        return .queen
    case "J":
        return .jack
    default:
        return Value(rawValue: Int(valueText)! - 2)!
    }
}

func parseSuit(_ suitText:String) -> Suit {
    if suitText == "C" {
        return .club
    } else if suitText == "S" {
        return .spade
    } else if suitText == "H" {
        return .heart
    } else {
        return .diamond
    }
}
